.circleci/parse_yaml.py

```diff
@@ -0,0 +1,273 @@
+    and trigger cimr processing of the contributed file
+    """
+
+    def __init__(self, yaml_data):
+        self.yaml_data = yaml_data
+        self.data_type = None
+        self.keys = None
+        self.hash = None
+        self.outdir = None
+        self.downloaded_file = None
+
+
+    def pick_keys(self):
+        """List keys for the dictionarized yaml data and store in self.
+        The following keys are expected:
+        ['defined_as', 'data_file', 'contributor', 'data_info', 'method']
+        """
+        self.keys = self.yaml_data.keys()
+
+
+    def set_data_type(self):
+        """One of the following data_type variables are expected:
+        ['gwas', 'twas', 'eqtl', 'sqtl', 'pqtl', 'tad']
+        """
+        if self.yaml_data['data_info']['data_type'] is not None:
+            self.data_type = self.yaml_data['data_info']['data_type']
+        else:
+            logging.error(f' there is no data_type indicated.')
+            sys.exit(1)
+
+
+    def download(self):
+        """Check if provided weblink to the file exists. 
+        Download if verified.
+        """
+        path = self.yaml_data['data_file']['location']['url']
+        self.downloaded_file = path.split('/')[-1]
+
+        outdir_root = 'submitted_data/'
+        pathlib.Path(outdir_root).mkdir(exist_ok=True)
+        self.outdir = outdir_root + str(self.data_type) + '/'
+        pathlib.Path(self.outdir).mkdir(exist_ok=True)
+
+        if verify_weblink(path):
+            logging.info(f' starting download')
+            download_file(path, self.outdir)
+            self.hash = self.yaml_data['data_file']['location']['md5']
+        else:
+            logging.error(f' file unavailable')
+            sys.exit(1)
+
+
+    def bulk_download(self):
+        """Bulk download option assumes one of the following file types:
+        ['tgz', 'tar.gz', 'tar.bz2', 'tar.xz']
+        """
+        import os
+        import tarfile
+
+        self.download()
+
+        if tarfile.is_tarfile(self.downloaded_file):
+            tarred_data = tarfile.open(
+                self.downloaded_file, 
+                mode='r:*'
+            )
+            for member in tarred_data.getmembers():
+                if member.isreg():
+                    member.name = os.path.basename(member.name)
+                    tarred_data.extract(member, path=self.outdir)
+    
+
+    def check_hash(self):
+        """Compare md5 of the downloaded file to the provided value"""
+        if validate_hash(self.downloaded_file, self.hash):
+            logging.info(f' data is ready for cimr processing.')
+            return True
+        else:
+            logging.error(f' provided md5 hash didn\'t match the downloaded file.')
+            return False
+
+
+    def check_defined(self):
+        import os
+
+        """Check whether the submitted data is a single file"""
+        if self.yaml_data['defined_as'] == 'upload':
+            self.download()
+            self.check_hash()
```

Comment:
If `check_hash()` returns a bool (as in current implementation) instead of raising an exception (as I proposed), then we have to check its return value (as you did on lines 241-242).